### PR TITLE
implement support for FTP/SFTP server

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
implement support for FTP/SFTP server

## Motivation and Context
(S)FTP server designed to use the handler
layer, the intention is to make sure
that (S)FTP server honors all the S3 
requirements including

- SSE (server-side encryption)
- Replication (bucket/site replication)

Passwords are to be validated again via 
signature v4. Credentials also honor
IAM policies - existing permissions 
automatically apply.

## How to test this PR?
Documentation is available at `docs/ftp/README.md` 

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [x] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/ftp-server-driver.go [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+func (m *minioFileInfo) Name() string {
+	return m.p
+}
+
+func (m *minioFileInfo) Size() int64 {
+	return m.info.Size
+}
+
+func (m *minioFileInfo) Mode() os.FileMode {
+	return os.ModePerm
```
Comment (klauspost) [inline cmd/ftp-server-driver.go:81]: I can't find the mc metadata key we had at one point. If we still have it, should we preserve this as a metadata field?

File: cmd/ftp-server-driver.go [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+	if err != nil {
+		return err
+	}
+
+	cctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	bucket, prefix := path2BucketObject(path)
+	if bucket == "" {
+		// TODO implement bucket listing.
```
Comment (klauspost) [inline cmd/ftp-server-driver.go]: Don't we pretty much need this?

File: cmd/ftp-server.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	if serverDebugLog {
+		logger.Info("%s < %d %s", sessionID, code, message)
+	}
+}
+
+func startFTPServer(c *cli.Context) {
+	tls := c.String("ftp-tls-private-key") != "" &&
+		c.String("ftp-tls-public-cert") != ""
+
+	ftpServer, err := server.NewServer(&server.Options{
```
Comment (klauspost) [inline cmd/ftp-server.go]: We need the passive port range to be specified, so it doesn't just pick random ones.

Comment (klauspost) [inline cmd/ftp-server.go]: I looked for a way to completely disabling active connections, but didn't find any.

**Review by klauspost — COMMENTED** [review]
Nothing major with the implementation. Private overall issues.

File: cmd/ftp-server-driver.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/ftp-server-driver.go:81]: Well, we do but they are not really meaningful here - on the server that's why these are just some dummy values that are okay.

The primary use case here is not full-blown FTP implementation just implementing it enough to satisfy, PUT/GET/LIST

File: cmd/ftp-server-driver.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/ftp-server-driver.go]: yes forgot to push the change

File: cmd/ftp-server.go (hunk 3 above)
Comment (harshavardhana) [reply cmd/ftp-server.go]: Will add

File: cmd/ftp-server-driver.go [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+		for object := range clnt.ListObjects(cctx, bucket, opts) {
+			if object.Err != nil {
+				return
+			}
+			objectsCh <- object
+		}
+	}()
+
+	// Call RemoveObjects API
+	for err := range clnt.RemoveObjects(context.Background(), "my-bucketname", objectsCh, minio.RemoveObjectsOptions{}) {
```
Comment (poornas) [inline cmd/ftp-server-driver.go]:
```suggestion
	for err := range clnt.RemoveObjects(context.Background(), bucket, objectsCh, minio.RemoveObjectsOptions{}) {
```

**Review by poornas — CHANGES_REQUESTED** [review]

File: cmd/ftp-server-driver.go [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+	}
+
+	dirPath := buildMinioDir(prefix)
+
+	_, err = clnt.PutObject(context.Background(), bucket, dirPath, bytes.NewReader([]byte("")), 0,
+		// Always send Content-MD5 to succeed with bucket with
+		// locking enabled. There is no performance hit since
+		// this is always an empty object
+		minio.PutObjectOptions{SendContentMd5: true},
+	)
```
Comment (poornas) [inline cmd/ftp-server-driver.go:437]: if bucket name does not exist, should this create a bucket?

Comment (harshavardhana) [reply cmd/ftp-server-driver.go:437]: Currently assuming buckets are created prior, but MakeBucket can be called in this situation.

File: cmd/ftp-server.go (hunk 3 above)
Comment (harshavardhana) [reply cmd/ftp-server.go]: there is a way @klauspost we simply need to disable PORT cmd.

**Comment by klauspost** [comment]
For the regular FTP. How do you control the passive port range?

**Comment by harshavardhana** [comment]
> For the regular FTP. How do you control the passive port range?

This library simply uses Linux features where OS can give you a free-port whatever is available, if needed we can use a range that is just for us. I don't know if we need it, 
I might as well hide the flag and use it only in specific situations. @klauspost

File: cmd/sftp-server-driver.go [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+			}}, nil
+		}
+
+		bucket, object := path2BucketObject(r.Filepath)
+		if bucket == "" {
+			return nil, errors.New("bucket name cannot be empty")
+		}
+
+		if object == "" {
+			_, err := clnt.BucketExists(context.Background(), bucket)
```
Comment (poornas) [inline cmd/sftp-server-driver.go]:
```suggestion
			ok, err := clnt.BucketExists(context.Background(), bucket)
```

**Comment by klauspost** [comment]
> This library simply uses Linux features where OS can give you a free-port whatever is available, if needed we can use a range that is just for us. 

For everything not running bare metal, you'd need a port range to expose the ports. Even bare metal users will probably appreciate that they can specify a port range in firewalls.

**Review by poornas — APPROVED** [review]

**Review by klauspost — APPROVED** [review]

**Pull request merged**